### PR TITLE
vm: add RemoveBreakPoint support

### PR DESCRIPTION
**Pull request by ixje**
### Problem

https://github.com/nspcc-dev/neo-go/issues/3673

### Solution

add function

### CODE REVIEW

**Review by AnnaShaleva — APPROVED** [review]
LGTM.

File: pkg/vm/vm.go
Comment (AnnaShaleva) [inline pkg/vm/vm.go:279]: BTW, can you also add a support of respective VM CLI command? Similar to https://github.com/nspcc-dev/neo-go/blob/66fbcb2f0032ae26350bf5b07231419e2195c8d5/cli/vm/cli.go#L104

If not, then let's move it to a separate issue.

**Comment by AnnaShaleva** [comment]
`Contribution guidelines` workflow is failing, could you please add a `signed-off-by` line to the commit message? Ref. https://github.com/nspcc-dev/.github/blob/master/git.md#commit-messages.

File: pkg/vm/vm.go
Comment (ixje) [reply pkg/vm/vm.go:279]: should I also add a `list breakpoints`?

File: cli/vm/cli.go
Comment (AnnaShaleva) [inline cli/vm/cli.go:118]: Useless change.

File: pkg/vm/vm.go
Comment (AnnaShaleva) [reply pkg/vm/vm.go:279]: To me it's a useful feature. It may be supported as a separate command and may be even as an extension of `(v *VM) PrintOps` (some special symbol may be added to every output line that has a breakpoint).

But keep in mind that `step` command is also handled with the help of breakpoints:
https://github.com/nspcc-dev/neo-go/blob/66fbcb2f0032ae26350bf5b07231419e2195c8d5/cli/vm/cli.go#L1191

Hence, the output of `list breakpoints` contains not only user-defined-breakpoints. This fact should be mentioned in the command doc.

Comment (AnnaShaleva) [reply pkg/vm/vm.go:279]: Another note is that we now actually have a set of breakpoint-related commands, hence we may group them under a single `break` command as a set of subcommands, something like: `break add`, `break remove` and `break list`:
```
var commands = []*cli.Command{
	{
		Name: "break",
		Subcommands: []*cli.Command{
			{
				Name: "break",
				...
			},
			{
				Name: "remove",
				...
			},
			{
				Name: "list",
				...
			},
		},
	},
```
But it's an incompatible change. @roman-khimov, what do you think?

File: cli/vm/cli.go
Comment (AnnaShaleva) [inline cli/vm/cli.go:609]: It requires a unit-test, see `TestRun_WithNewVMContextAndBreakpoints` for reference.

File: pkg/vm/vm.go
Comment (roman-khimov) [reply pkg/vm/vm.go:279]: I think `man gdb`.

Comment (AnnaShaleva) [reply pkg/vm/vm.go:279]: Well, I had to be more clear: what do you think about *incompatibility* of this change, can we afford this?

Regarding `gdb`:  it has the same concept as proposed above: `gdb breakpoints` has a subset of breakpoint-related commands including `break`, `clear`, `delete`. Although I haven't find something similar to `list` to display all breakpoints.

Comment (ixje) [reply pkg/vm/vm.go:279]: `ib` lists breakpoints in gdb

**Comment by ixje** [comment]
Slightly off-topic but relevant for me. Are you open to exposing the following?
https://github.com/nspcc-dev/neo-go/blob/dda2cafdf8f7a30905bb0b8516b41ec20dce2c21/pkg/vm/context.go#L33
and
https://github.com/nspcc-dev/neo-go/blob/dda2cafdf8f7a30905bb0b8516b41ec20dce2c21/pkg/vm/context.go#L68-L69

because if not then I'll have to fork anyway and we can skip this PR and #3675 to save you some time. I need access to these slots or I can't display stack frame information. I'm aware of the `dumpSlot()` output but that has some issues and I'd prefer to transfer the information to the IDE in a binary format

File: pkg/vm/vm.go
Comment (roman-khimov) [reply pkg/vm/vm.go:279]: > incompatibility of this change

If you really want to debug something with this CLI you won't be happy typing "breakpoints break" and alike, these:
```
break, brea, bre, br, b -- Set breakpoint at specified location.
delete, del, d -- Delete all or some breakpoints.
```
are all available at the top level in GDB. Also, good commands start with verbs.

Strict compatibility is not critical for this CLI (and can be done with some transition period), but likely we can extend it in a compatible way now.

**Comment by AnnaShaleva** [comment]
> Are you open to exposing the following?

If an external application needs it, then I don't have objections. But `(c *Context) Dump*Slot` methods exist because `vm.slot` structure is unexported and I think we should not export it because it's an internal VM structure. Hence, I'd suggest to make a set of `func (c *Context) *SlotBytes() []byte` helpers in a separate commit that will serialize slots using binary representation instead of JSON.

> but that has some issues

Could you please clarify, which issues? Because I'd expect `dumpSlot` to output a proper JSON that may be reused by other applications.

> we can skip this PR and https://github.com/nspcc-dev/neo-go/issues/3675

Let's finalize it, I consider this PR useful.

File: pkg/vm/vm.go
Comment (AnnaShaleva) [reply pkg/vm/vm.go:279]: OK, let's then agree on compatible extension and use `delete` and `ib` for breakpoints removal and listing correspondingly.

**Comment by ixje** [comment]
> `vm.slot` structure is unexported and I think we should not export it because it's an internal VM structure. 

I understand this is sensitive data that shouldn't be changed, but how are they different from the current `Estack()` and `Istack()` methods? Those equally give you enough power to wreck the VM state.

> Hence, I'd suggest to make a set of `func (c *Context) *SlotBytes() []byte` helpers in a separate commit that will serialize slots using binary representation instead of JSON.

I'll consume whatever is made available :pray: 

> Could you please clarify, which issues? Because I'd expect `dumpSlot` to output a proper JSON that may be reused by other applications.

Assume the following
```golang
type DebugContext {
   StackFrames []StackFrame  `json:"stackFrames"`
}

type StackFrame struct {
	Arguments string `json:"arguments"`
	Statics   string `json:"statics"`
	Locals    string `json:"locals"`
	NextIP    int    `json:"nextIp"`
	CurIP     int    `json:"curIp"`
}
...
	for i, ctx := range ic.VM.Istack() {
		frames[i] = StackFrame{
			Arguments: ctx.DumpArgumentsSlot(),
			Statics:   ctx.DumpStaticSlot(),
			Locals:    ctx.DumpLocalSlot(),
			NextIP:    ctx.NextIP(),
			CurIP:     ctx.IP(),
		}
	}
```
marshalling this gives for example
```
{"stackFrames":[{"arguments":"[\n    {\n        \"type\": \"Integer\",\n        \"value\": \"1\"\n    }\n]","statics":"[]","locals":"[\n    null,\n    null\n]","nextIp":3,"curIp":0}]}
```
1 problem here is that `arguments` (as are `statics` & `locals`) is a string not an array. So to get it into a usable format I have 2 choices
1. on the neo-go side, unmarshal the dumpSlot returned string into a slice of some new type. Then when the whole `DebugContext` is marshaled `arguments` will be a proper array
2. on the consumer side read the value of e.g. `arguments`, then run another JSON parser instance on the value.

**Comment by AnnaShaleva** [comment]
> but how are they different from the current Estack() and Istack() methods? 

From that point you're right, so to be honest I don't have any strong objections against of creating a set of `func (c *Context) *Slot() Slot` methods. Let's implement this.

**Comment by ixje** [comment]
I can't seem to reply to https://github.com/nspcc-dev/neo-go/pull/3674#discussion_r1838090081 so I'll do it here

> Hence, the output of list breakpoints contains not only user-defined-breakpoints. This fact should be mentioned in the command doc.

I did not add a mention because I believe the current implementation of `step n` is flawed (see 
https://github.com/nspcc-dev/neo-go/issues/3676). I expect the solution to not use a break point, therefore it won't show up.

**Review by AnnaShaleva — COMMENTED** [review]
Please, adjust the commit message to describe all changes implemented in this commit or split the RP into multiple commits (ref. https://github.com/nspcc-dev/.github/blob/master/git.md#logical-separation). Also, `Contribution guidelines` and `Lint` jobs are failing, add `signed-off-by` to the commit message and fix the linter issues.

Otherwise LGTM.

**Comment by ixje** [comment]
image

:thinking:

**Comment by ixje** [comment]
if this ain't the right format then I give up

**Comment by AnnaShaleva** [comment]
> if this ain't the right format

It's the right one!

**Review by AnnaShaleva — APPROVED** [review]
Also needs @roman-khimov's approval.

**Comment by AnnaShaleva** [comment]
BTW, please also add `Close #3673.` to the commit message to auto-close related GH issue after merge.

**Review by roman-khimov — APPROVED** [review]
I'd really prefer to see real names in Author/SOB (and log, respectively), but it's up to you.

**Pull request merged**
